katcp/resource.py

```diff
@@ -350,8 +383,27 @@ def __call__(self, *args, **kwargs):
         reply : tornado future resolving with :class:`KATCPReply` object
             KATCP request reply wrapped in KATCPReply object
 
+        Raises
+        ------
+        :class:`ResourceInactive` if the resource is inactive when the request is made.
+        """
+        if self.is_active():
+            return self.issue_request(*args, **kwargs)
+        else:
+            raise KATCPResourceInactive(
+                "Can't make ?{} request; resource is inactive".format(self.name))
+
+    @abc.abstractmethod
+    def issue_request(self, *args, **kwargs):
+        """Signature as for __call__
+
+        Do the request immediately without checking active state.
         """
 
+    def is_active(self):
+        """True if resource for this request is active"""
+        return self._is_active()
```

Comment:
Indeed not a method, but it is a callable member that is set by __init__() unless I'm missing something.

The design-idea is that KATCPRequest doesn't itself have any idea of how 'activity' is determined, but trusts the callable passed in at __init__ time.